### ISSUE TITLE
Best practice for multi-tenant cluster?

### ISSUE THREAD

**Issue by jhyot** [Azure/aad-pod-identity#1163]
What is the recommended security practice regarding aad-pod-identity for an AKS cluster where multiple "tenants" (or independent teams) are deploying their applications?

The individual tenants obviously only have read/write permissions to their own namespace. But even if aad-pod-identity is running in namespaced mode, anybody who can create an `AzureIdentity` and `AzureIdentityBinding` in their namespace, can bind their pod to any Azure managed identity that is being used somewhere else on the cluster.

Example: Team A is working in Kubernetes namespace-a, and has created a Managed Identity "Team A Identity" that has read access to Team A's Key Vault.
Team B, in namespace-b, can now create their own `AzureIdentity` and `AzureIdentityBinding` to "Team A Identity", and so access the other team's Key Vault.

Forbidding creation of new AzureIdentities for "normal" tenants might work, but we also don't want to have a manual admin process to create new namespaces and identities for individual teams. How could this be handled in a secure and automated way?

Ideally there would be a restriction on which Managed Identities can be bound into a namespace, e.g.: "in namespace-b, only AzureIdentities from Resource Group 'team-b' can be created and bound".

**Comment by chewong** [comment]
According to your example, if team B has the resource ID and the client ID of team A identity (they can get the information via azure-cli or Azure portal), team B can create an AzureIdentity based on team A identity.

> Ideally there would be a restriction on which Managed Identities can be bound into a namespace, e.g.: "in namespace-b, only AzureIdentities from Resource Group 'team-b' can be created and bound".

This is out of the project's scope. I think this can be achieved by revoking read access from team B on team A identity via RBAC on Azure.

**Comment by jhyot** [comment]
> According to your example, if team B has the resource ID and the client ID of team A identity (they can get the information via azure-cli or Azure portal), team B can create an AzureIdentity based on team A identity.

Yes that's what I mean. As far as I have understood it, resource and client IDs are not *secrets* per se (they are IDs), so usually no effort is done to specifically conceal or hide these IDs. There could be potentially many ways for a third party to obtain these IDs.
 
> > Ideally there would be a restriction on which Managed Identities can be bound into a namespace, e.g.: "in namespace-b, only AzureIdentities from Resource Group 'team-b' can be created and bound".
> 
> This is out of the project's scope. I think this can be achieved by revoking read access from team B on team A identity via RBAC on Azure.

This would prevent for team B to read the IDs in the most direct way (by reading the resource in Azure), but this would not prevent team B from creating a binding to team A's identity in AKS once they have the IDs by other means (e.g. from logs, checked in into public repositories, by accident, ....), would it?

So do you (or anybody else) have a comment how to handle this scenario in the best, secure, way? I'm feeling I can't be the first one to think of this problem.

I'm especially interested how to handle this in an automated way. Of course there could be an admin who has the sole rights to create AzureIdentitites, and the teams could email this person whenever they need it. But I feel that Kubernetes is part of the "more modern" world of automation and self-provisioning, and the email solution does not scale.

Or is AKS maybe not really recommended for a "no-trust" multi-tenant cluster setup? I can live with that answer too (and maybe raise it with the AKS team for discussion).

**Comment by chewong** [comment]
> So do you (or anybody else) have a comment how to handle this scenario in the best, secure, way? I'm feeling I can't be the first one to think of this problem.

It's totally up to the cluster admin on how to handle this and we don't have a best practice for it at the moment. IMO, you could
- If possible, create two separate AKS clusters for team A and team B. You could also create them in two separate Azure subscriptions for additional boundary.
- Leverage [gatekeeper](https://github.com/open-policy-agent/gatekeeper) to block the creation of AzureIdentity with certain specs defined within them (e.g. block the creation of team A's AzureIdentities outside of team A's namespace). You can check out our example policy - https://github.com/Azure/aad-pod-identity/blob/master/validation/gatekeeper/azureidentityformat_template.yaml

**Comment by mblaschke** [comment]
I've wrote a small operator which maintains `AzureIdentity` and `AzureIdentityBinding` in a secure way:

https://github.com/webdevops/azure-msi-operator
